elkscmd/unused/mtools/Mmd.1
.TH MMD 1 local
.SH NAME
mmd \- make a MSDOS sub directory
.SH SYNOPSIS
.B mmd
[
.B -v
] msdosdirectory
.SH DESCRIPTION
.I Mmd
makes a new directory on a MSDOS formatted diskette.  Sub directories
are supported with either the '/' or '\e\' separator.  The use of
the '\e\' separator will require the directory name to be enclosed in
quotes to protect it from the shell.
.PP
.I Mmd
will allow the following command line option:
.TP
.B v
Verbose mode.  Display the new directory name if the name supplied is
invalid.
.PP
Reasonable care is taken to create a valid MSDOS directory name.  If an
invalid name is specified,
.I mmd
will change the name (and display the new name if the verbose mode is
set).
.PP
An error occurs if the directory already exists.
.PP
The environmental variable MCWD may be used to establish a current
working directory (relative to the MSDOS diskette).
.SH SEE ALSO
mdir(1), mrd(1)


elks/CHANGELOG
This file is used to record changes made to the CVS database. Please add to
the top of this file before committing any changes. Please use the header
format provided for each entry, noting that the text on each line should not
be longer than the lines of dashes.

Al Riddoch.

Contributor           Email address                     CHANGELOG entries found
~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~
Alan Cox              alan@redhat.com                   17 Jun 2001-18 Jun 2001
Alastair Bridgewater  nyef@softhome.net                 19 Nov 1998- 6 Dec 1998
Alistair Riddoch      alriddoch@zepler.org              25 Sep 1998-17 Jun 2001
Arnaud Launay         asl@launay.org                    15 Jul 1999-17 Oct 2001
Christian Mardmuller  chm@kdt.de                        10 Nov 1999-29 Feb 2000
Christoph Plattner    Christoph.Plattner@dot.at         31 Jan 2000- 2 Feb 2000
Claudio Matsuoka      claudio@conectiva.com
Daniele Giacomini     daniele@swlibero.org		20 Sep 2004
David Murn            scuffer@hups.apana.org.au         11 Nov 1998-15 Nov 1999
Eduardo Habkost	      ehabkost@conectiva.com.br		01 Jun 2004
Giles Russell                                           22 Nov 1999
Greg Haerr                                              14 Jul 1999-20 Jul 1999
Harry Kalogirou       HarKal@gmx.net                    20 Jun 2001- 2 Jan 2003
Jacek Lipkowski       sq5bpf@acid.ch.pw.edu.pl          20 Jan 2002
Jay Barr                                                27 Oct 2002
Ken Martwick          kenm@efn.org                      12 May 2002
M Warner Losh         imp@village.org                   20 Jan 2002
Manuel Novoa III      mjn3@codepoet.org                 18 Jun 2002-25 Jun 2002
Mario Frasca                                            20 Jan 2000-14 Feb 2000
Miguel Bolanos	      mike@linuxlabs.com		20 Sep 2004
Paul Nasrat           pauln@truemesh.com                29 Oct 2001-12 Jul 2002
Per Olofsson          MagerValp@cling.gu.se             25 Feb 2000
Peter L Vachuska      pvachusk@nconnect.net             13 Aug 2001
Richard Wallman       murkygoth@murkygoth.uklinux.net    2 Apr 2002
Riley Williams        Riley@Williams.Name               16 Nov 1998-13 Jul 2003
Roman Festchook       roma@polesye.net                  27 Oct 2001
Simon Weijgers        simon@weijgers.com                16 Sep 1998-15 Nov 1999
Simon Wood                                               6 Jan 2000-12 Jan 2000
Thomas McWilliams     tgm9999@yahoo.com                 27 Aug 1999- 2 Apr 2001
Tommy McCabe 	      rocketjet314@yahoo.com		31 May 2004

-------------------------------------------------------------------------------

Thu Feb 16 17:03:00 EST 2012	Jody Bruchon <jody@jodybruchon.com>	
	* Removed elksfs. It was a copy of the MINIX fs and no tools
	  exist to make such a filesystem.

Thu Nov 23 21:35:28 EST 2006	Segin <segin2005@gmail.com>
	* Changed name for the submenu for selecting charater
	  device drivers to note the fact that that submenu is broken;
	  attempting to enter it will have lxdialog die horribly.

Wed Nov  8 17:49:52 EST 2006	Segin <segin2005@gmail.com>

	* Changed ELKS init sequence to check all locations for 'init'
	  that the Linux kernel does. This is to follow more 
	  "traditional" behaviour.

Sun Aug 20 21:51:25 UTC 2006	Claudio Matsuoka <claudio@mandriva.com>

	* Allow loading Minix executables with total size larger than
	  64KB, such as ash (70140 bytes). Currently the total size
	  minus 32 bytes of header was limited to 16-bit.

	* Set tty1 back to minor 0. This is different from Linux
	  numbering, but it's necessary to have current ELKS virtual
	  terminals working. This and the executable size change put
	  ELKS back in usable state, allowing logins in the first VT
	  and executing a shell.

Mon Sep 20 23:27:23 CST 2004	Daniele Giacomini <daniele@swlibero.org>

	* Added italian keys support, Miguel Bolanos cleaned the original
	  file code.

Wed Jun  9 17:27:52 BRT 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* fix for my last fix: s/rp_cmd/rq_cmd/

Thu Jun 03 21:47:13 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* fixed format specifiers to correctly displaysegment sizes as unsigned
	  values, and to display debug messages correctly.

Thu Jun 03 07:48:25 CST 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* Fixed wrong struct field name for arch/i86/drivers/block/doshd.c
	* Several fixes ro arch/i86/mm/user.c

Web Jun 02 14:16:57 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* Added an extra fix to Eduardos memcpy fix.

Tue Jun 01 21:00:03 CST 2004	Eduardo Habkost ehabkost@conectiva.com.br

	* Fix memcpy_{to,from}fs, adding code to save the contents of SI and DI

Mon May 31 07:39:47 CST 2004	Tommy McCabe <rocketjet314@yahoo.com>

	* Fix wrong assignment (typo?) on printk
	* Fix order of LD parameters on arch/i86/Makefile
	* Use #ifdef rightly on setup.S
	* Fix init to use tty0, not tty1 as system console.

===============================================================================

Sun Jul 13 04:24:44 BST 2003	Riley Williams <Riley@Williams.Name>

	* This is ELKS 0.1.3-pre1

Sat Jul 12 12:24:36 BST 2003	Riley Williams <Riley@Williams.Name>

	* Rewrote kernel/printk.c to allow printk() to correctly print out
	  signed values.
	* Rewrote kernel/printk.c to produce a less verbose panic() dump,
	  and also to increase the number of lines of stack trace to 10.
	* Various spelling corrections to printk() messages.

===============================================================================

Thu Jan 02 16:12:36 EET 2003	Harry Kalogirou <harkal@gmx.net>

        * Better select hash.
	* This is ELKS 0.1.2

-------------------------------------------------------------------------------

Fri Nov 01 01:52:44 EET 2002	Harry Kalogirou <harkal@gmx.net>

        * Changed the position of ds in the _registers struct.
        * Added support for loading the new ELKS executable format. This is 

elkscmd/mtools/mmd.c
/*
 * Make a MSDOS sub directory
 *
 * Emmet P. Gray			US Army, HQ III Corps & Fort Hood
 * ...!uunet!uiucuxc!fthood!egray	Attn: AFZF-DE-ENV
 * 					Directorate of Engineering & Housing
 * 					Environmental Management Office
 * 					Fort Hood, TX 76544-5057
 */

#include <stdio.h>
#include "msdos.h"

int fd;				/* the file descriptor for the floppy */
int dir_start;			/* starting sector for directory */
int dir_len;			/* length of directory (in sectors) */
int dir_entries;		/* number of directory entries */
int dir_chain[25];		/* chain of sectors in directory */
int clus_size;			/* cluster size (in sectors) */
int fat_len;			/* length of FAT table (in sectors) */
int num_clus;			/* number of available clusters */
unsigned char *fatbuf;		/* the File Allocation Table */
char *mcwd;			/* the Current Working Directory */

main(argc, argv)
int argc;
char *argv[];
{
	int entry, slot, fat, dot, fargn, verbose;
	char *filename, *newfile, *fixname(), *strncpy();
	char *getpath(), *pathname, tname[9], text[4], *fixed, *unixname();
	void exit(), putcluster(), writefat(), writedir(), free();
	struct directory *dir, *search(), *mk_entry();
	long time(), now;

	if (init(2)) {
		fprintf(stderr, "mmd: Cannot initialize diskette\n");
		exit(1);
	}

	fargn = 1;
	verbose = 0;
	if (argc > 1) {
		if (!strcmp(argv[1], "-v")) {
			fargn = 2;
			verbose = 1;
		}
	}
					/* only 1 directory ! */
	if (argc != fargn+1) {
		fprintf(stderr, "Usage: mmd [-v] msdosdirectory\n");
		exit(1);
	}
	fixed = fixname(argv[fargn], verbose);

	strncpy(tname, fixed, 8);
	strncpy(text, fixed+8, 3);
	tname[8] = '\0';
	text[3] = '\0';

	filename = unixname(tname, text);
	pathname = getpath(argv[fargn]);

	if (subdir(pathname))
		exit(1);
					/* see if exists and get slot */
	slot = -1;
	dot = 0;
	for (entry=0; entry<dir_entries; entry++) {
		dir = search(entry);
					/* if empty */
		if (dir->name[0] == 0x0) {
			if (slot < 0)
				slot = entry;
			break;
		}
					/* if erased */
		if (dir->name[0] == 0xe5) {
			if (slot < 0)
				slot = entry;
			continue;
		}
					/* if not a directory */
		if (!(dir->attr & 0x10))
			continue;

		strncpy(tname, (char *) dir->name, 8);
		strncpy(text, (char *) dir->ext, 3);
		tname[8] = '\0';
		text[3] = '\0';

		newfile = unixname(tname, text);
					/* save the 'dot' directory info */
		if (!strcmp(".", newfile))
			dot = dir->start[1]*0x100 + dir->start[0];

		if (!strcmp(filename, newfile)) {
			fprintf(stderr, "mmd: Directory \"%s\" already exists\n", filename);
			exit(1);
		}
		free(newfile);
	}
					/* no '.' entry means root directory */
	if (dot == 0 && slot < 0) {
		fprintf(stderr, "mmd: No directory slots\n");
		exit(1);
	}
					/* make the directory grow */
	if (dot && slot < 0) {
		if (grow(dot)) {
			fprintf(stderr, "mmd: Disk full\n");
			exit(1);
		}
					/* first slot in 'new' directory */
		slot = entry;
	}
	if ((fat = nextfat(0)) == -1) {
		fprintf(stderr, "mmd: Disk full\n");
		exit(1);
	}
					/* make directory entry */
	time(&now);
	dir = mk_entry(fixed, 0x10, fat, 0L, now);
	writedir(slot, dir);
					/* write the cluster */
	putfat(fat, 0xfff);
	putcluster(fat, dot);
					/* write FAT sectors */
	writefat();
	close(fd);
	exit(0);
}

/*
 * Write a blank directory 'template' to the cluster starting at 'dot'.
 */

void
putcluster(dot, dot_dot)
int dot, dot_dot;
{
	int buflen, start;
	static struct directory dirs[32];
	struct directory *mk_entry();
	void exit(), perror(), move();
	long time(), now;

	start = (dot - 2)*clus_size + dir_start + dir_len;
	move(start);

	buflen = clus_size * MSECSIZ;
					/* make the '.' and '..' entries */	
	time(&now);
	dirs[0] = *mk_entry(".          ", 0x10, dot, 0L, now);
	dirs[1] = *mk_entry("..         ", 0x10, dot_dot, 0L, now);

	if (write(fd, (char *) &dirs[0], buflen) != buflen) {
		perror("putcluster: write");
		exit(1);
	}
	return;
}

/*
 * Returns next free cluster or -1 if none are available.
 */

int
nextfat(last)
int last;
{
	register int i;

	for (i=last+1; i<num_clus+2; i++) {
		if (!getfat(i))
			return(i);
	}
	return(-1);
}
